src/utils/DirectHub.cxx

```diff
@@ -0,0 +1,943 @@
+    /// Type of the queue used to keep the output buffer queue.
+    typedef Q QueueType;
+
+    /// total number of bytes written to the port.
+    size_t totalWritten_ {0};
+
+    /// The buffer that is taken out of the queue while flushing.
+    BufferPtr<OutputDataEntry> currentHead_;
+    /// Data we are currently writing to a buffer.
+    DataBuffer *nextToWrite_;
+    /// Skip_ parameter matching nextToWrite_;
+    unsigned nextToSkip_;
+    /// Size_ parameter matching nextToWrite_;
+    unsigned nextToSize_;
+    /// Helper object for performing asynchronous writes.
+    StateFlowSelectHelper selectHelper_ {this};
+    /// Time when the last buffer flush has happened. Not used yet.
+    // long long lastWriteTimeNsec_ = 0;
+
+    /// Contains buffers of OutputDataEntries to write.
+    QueueType pendingQueue_;
+    /// Last tail pointer in the pendingQueue. If queue is empty,
+    /// nullptr. Protected by pendingQueue_.lock().
+    OutputDataEntry *pendingTail_ = nullptr;
+    /// Total numberof bytes in the pendingQueue.
+    size_t totalPendingSize_ = 0;
+    /// 1 if the state flow is paused, waiting for the notification.
+    uint8_t notRunning_ : 1;
+    /// 1 if the read flow is still running.
+    uint8_t readFlowPending_;
+    /// 1 if the write flow is still running.
+    uint8_t writeFlowPending_;
+    /// Parent hub where output data is coming from.
+    DirectHubInterface<uint8_t[]> *hub_;
+    /// File descriptor for input/output.
+    int fd_;
+    /// This notifiable will be called before exiting.
+    Notifiable *onError_ = nullptr;
+};
+
+extern DirectHubPortSelect *g_last_direct_hub_port;
+DirectHubPortSelect *g_last_direct_hub_port = nullptr;
+
+void create_port_for_fd(DirectHubInterface<uint8_t[]> *hub, int fd,
+    std::unique_ptr<MessageSegmenter> segmenter, Notifiable *on_error)
+{
+    g_last_direct_hub_port =
+        new DirectHubPortSelect(hub, fd, std::move(segmenter), on_error);
+}
+
+class DirectGcTcpHub
+{
+public:
+    /// Constructor.
+    ///
+    /// @param can_hub Which CAN-hub should we attach the TCP gridconnect hub
+    /// onto.
+    /// @param port TCp port number to listen on.
+    DirectGcTcpHub(DirectHubInterface<uint8_t[]> *gc_hub, int port);
+    ~DirectGcTcpHub();
+
+    /// @return true of the listener is ready to accept incoming connections.
+    bool is_started()
+    {
+        return tcpListener_.is_started();
+    }
+
+private:
+    /// Callback when a new connection arrives.
+    ///
+    /// @param fd filedes of the freshly established incoming connection.
+    ///
+    void OnNewConnection(int fd);
+
+    /// Direct GridConnect hub.
+    DirectHubInterface<uint8_t[]> *gcHub_;
+    /// Helper object representing the listening on the socket.
+    SocketListener tcpListener_;
+};
+
+void DirectGcTcpHub::OnNewConnection(int fd)
+{
+    uint32_t rcvbuf;
+    socklen_t len = sizeof(rcvbuf);
+    int ret = getsockopt(fd, SOL_SOCKET, SO_RCVBUF, &rcvbuf, &len);
```

Comment:
this call appears to be for debug / diagnostic, perhaps it should be wrapped / excluded from production code? Or perhaps the value used by the port instance to limit data?

Reply:
done